### ISSUE TITLE
Refactor vertex functions to enable composite paths

### ISSUE THREAD

**Issue by GregStanton** [processing/p5.js#6560]
### Increasing Access

This enhancement will provide greater access to individuals in multiple circumstances.

**Code contributors:**
For volunteers who have less experience with large codebases or less free time to navigate them, it will become feasible to contribute to vertex-related features. For example, there is community support for a new `arcVertex()` function, but the complexity of the current codebase has impeded progress.

**Users:**
For users who want to make composite shapes, the existing p5 API will work as they expect it to, so they will face less confusion. A knowledge barrier will also be removed, since they won't need to learn the native canvas API or SVG in order to make composite shapes.

### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [X] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Build Process
- [ ] Unit Testing
- [ ] Internalization
- [ ] Friendly Errors
- [ ] Other (specify if possible)

#### Mixing path primitives
Users expect to be able to create composite paths with a mix of path primitives (line segments, quadratic Bézier curves, cubic Bézier curves, and Catmull-Rom splines). The p5 reference does not indicate that general mixing is not possible, and it's already possible with native canvas paths and SVG paths. 

However, the current p5 implementation assumes shapes include at most one type of nonlinear path. For example, see the [2D renderer's `endShape()` method](https://github.com/processing/p5.js/blob/v1.8.0/src/core/p5.Renderer2D.js#L835). This limitation leads to unexpected behavior in a range of cases. A couple of these are illustrated below.

| | Expected shape | Actual result |
| :--- | :--- | :--- |
| **Quadratic and cubic Béziers** | <img src = "https://github.com/processing/p5.js/assets/51820777/f342de37-719b-4631-a3a7-0bf383c58515" width = "150">| <img src = "https://github.com/processing/p5.js/assets/51820777/7943cec3-0a6b-463d-ba08-ad6bcad43878" width = "150">|
| **Lines and Catmull-Rom splines** | <img src = "https://github.com/processing/p5.js/assets/51820777/3b83a5d8-4920-4b4f-b909-e0056c793900" width = "150">  | <img src = "https://github.com/processing/p5.js/assets/51820777/80a2c467-e666-49d6-a640-b7cbd3d1a41a" width = "150">|

#### Complexity of the codebase
With the current implementation, it will be difficult to support mixing vertex types (e.g. #906), to track down bugs (e.g. #6555), and to implement new features (e.g. #6459). The difficulty arises from specific functions as well as broader code organization:

* The 2D renderer's `endShape()` method is an example of a specific function with room for improvement. It has too many responsibilities, as it contains rendering code for all vertex types. It also directly includes a [mathematical conversion from Catmull-Rom splines to Bezier curves](https://github.com/processing/p5.js/blob/v1.8.0/src/core/p5.Renderer2D.js#L863) that partly duplicates work done by [a separate function in p5.Font.js](https://github.com/processing/p5.js/blob/17304ce9e9ef3f967bd828102a51b62a2d39d4f4/src/typography/p5.Font.js#L1179), and it contains dead code (it repeatedly checks if shapeKind is equal to constants.POLYGON, but no such constant is defined in [constants.js](https://github.com/processing/p5.js/blob/main/src/core/constants.js)).
* The way code is divided across the codebase is an example of an organizational challenge. This [makes it difficult to determine where new drawing code should go](https://github.com/processing/p5.js/issues/6459#issuecomment-1793446406).

#### Separation of concerns
We can solve many of these problems at once by separating concerns. The idea is to create classes for the different vertex types, each with their own data and their own methods that add them to the canvas. Then `vertex()`, `quadraticVertex()`, etc. just push vertex instances of the appropriate type into a path array. When `endShape()` is called, it simply loops over the vertex objects, and they add themselves to the canvas; `endShape()` doesn't need to know that `curveVertex()` requires special logic, for example. By refactoring in this way, mixing path primitives is essentially automatic, with no change to the public p5 API.

#### Proof of concept
I put together [a proof of concept](https://editor.p5js.org/highermathnotes/sketches/R14UrggFq) in the p5 Web Editor; it includes an implementation of `arcVertex()` as proposed in issue #6459. [A separate sketch](https://editor.p5js.org/highermathnotes/sketches/JUxU5wx1D) shows how it fixes unexpected mixing behavior. With these two enhancements together (`arcVertex()` and composite paths), p5's `beginShape()`/`endShape()` will be able to produce composite paths with all the basic primitives of the native canvas API and SVG, plus Catmull-Rom splines. Based on some discussion, it looks like it will also help with the bug observed in issue #6555. Although the proof of concept focuses on the 2D case, the goal is to accommodate the 3D case as well, with some adjustments.

### Update: Task list
Calling all volunteers!

@davepagurek, @limzykenneth, @xanderjl, @Gaurav-1306, @Jaivignesh-afk, @sudhanshuv1, @capGoblin, @lindapaiste

Based on discussions here, in the online meeting, and on Discord, we're going to use [this comment](https://github.com/processing/p5.js/issues/6560#issuecomment-1844911703) as an overall implementation guide, and  we've decided to organize the work according to the following tasks:

- [ ] Update reference to explicitly describe the new behaviors
- [ ] Create snapshots of existing behavior, for the visual tests
- [ ] Decide how to refactor `normal()`*
- [ ] Decide whether to rename default contour kind to `SEGMENTS` or leave it as `null`*
- [ ] [Increase consistency in usage and documentation of default values](https://github.com/processing/p5.js/issues/6679)
- [ ] Decide on the file structure*
- [ ] Implement `Shape`, `Contour`, and primitive shape classes
- [ ] Implement creator functions*
    - [ ] Decide if we want wrappers (e.g. `createPoint()` for `() => new Point(\*data*\)`)
    - [ ] Implement wrappers or do nothing if we decide to omit them
- [ ] Implement `PrimitiveShapeCreators` (a JavaScript `Map` object)*
- [ ] Implement 2D and 3D output visitor classes
- [ ] Implement `drawShape()` on 2D and 3D renderers and refactor user-facing functions
- [ ] Develop `PointAtLengthGetter`**
- [ ] Consider new primitives based on existing combinations of vertex and shape kinds**

`*`: A separate GitHub Issue does not need to be created.
`**`: A separate GitHub Issue should be created, but not yet. 

For all the unmarked tasks, I'll set up separate issues with more details as soon as I get a chance, and the task list will link to them. Any code should be committed to the new [vertex-refactor](https://github.com/processing/p5.js/tree/vertex-refactor) branch @davepagurek set up, and we'll merge it into the main branch once the time is right. For right now, we can get started by discussing the tasks marked with a single asterisk!

**Comment by Gaurav-1306** [comment]
I would like to work on this issue.

Just a little hiccup, i can start in the next 2 weeks, so if someone wants to do it before that they can, but after that i would like to work on this issue. 
@GregStanton thanks again for the update 😃

**Comment by GregStanton** [comment]
Hi @Gaurav-1306! It's great that you want to work on this! We may want to break this up into smaller tasks that multiple people can work on. Maybe the first thing is to make a list of people who are interested. Maybe one person can take the lead by dividing up tasks. When I get a chance, I can think more about that, but I'd love to hear from others on this.

**Comment by Jaivignesh-afk** [comment]
I would like work on this as well!!

**Comment by limzykenneth** [comment]
Great write up @GregStanton! I like the idea of having internal classes for each type of vertices and it also makes it possible for addon libraries to implement their own unique vertices that can work with beginShape() and endShape(). 

As a start can we work out a common interface that each of the classes should adhere to? ie. What properties and methods should they all have? What are the signatures of the methods?

After that or at the same time we can figure out how the internals call the methods/interface. For anyone interested in working on this, please work out a proposal in this issue first. You can test out code implementation in the web editor as @GregStanton has done with the [proof of concept](https://editor.p5js.org/highermathnotes/sketches/R14UrggFq). Thanks!

**Comment by GregStanton** [comment]
Thanks for the kind words @limzykenneth, and for helping to direct the effort! I'm really glad you like the idea.

I'm not sure if you noticed, but the code that makes my proof of concept work is all located in shape.js, inside the project's file directory. I think it may provide initial answers to your questions. For example, the vertex classes all have the following interface: 

* `data` property
   * **description:** array holding the vertex and other data (e.g. control points) passed to the associated p5 function
   * **usage:** in the case of CurveVertex objects, this property holds data from consecutive calls to curveVertex()
* `type` property
   * **description:** type of vertex (possible values are constants.VERTEX, constants.QUADRATIC_VERTEX, etc.)
   * **usage:** curveVertex() function checks type to see if it needs to push data into an existing CurveVertex object
* `coordinates` getter
   * **description:** returns vertex coordinates from data (e.g. in CurveVertex, this is the second-to-last data point)
   * **usage:** endShape() uses it to keep track of the current point on the path, for use by ArcVertex 
* `addToCanvasPath()` method
   * **description:** takes no parameters; adds the corresponding subpath to the canvas
   * **usage:** called by endShape()

**Possible next steps/questions**
1.  Does anyone want to extend my proof of concept to the 3D case? This might mean adding a method or getter analogous to `addToCanvasPath()`, but it would output vertices; @davepagurek mentioned something along these lines on Discord.
2. Does anyone want to extend my proof of concept so that it supports contours? I think Dave also suggested a nested array that starts with _path as in shape.js and holds additional arrays for the contours after that.
3. I'm glad you mentioned extensibility; that's a good point to keep in mind. I guess we'd need to expose something like shape.js's _path array in order to allow add-on libraries to create new vertex types? Maybe there's another way.

**Comment by limzykenneth** [comment]
> I'm glad you mentioned extensibility; that's a good point to keep in mind. I guess we'd need to expose something like shape.js's _path array in order to allow add-on libraries to create new vertex types? Maybe there's another way.

We don't necessarily have to to figure this out at this point. Once we have the implementation working with the flexibility we want, we can revisit at a later point. That is to say it's a nice to have but not must have.

**Comment by davepagurek** [comment]
For 3D support: in WebGL, each path/contour ends up just as an array of vertices, so each segment would need to be able to convert itself into a polyline. This will also depend on the level of detail, so its interface for WebGL would have to look something like:
```ts
interface Segment {
  first: boolean
  addToCanvasPath(ctx: CanvasRenderingContext2D): void
  addToPolyline(polyline: Array<p5.Vector>, curveDetail: number): void
}
```

Here I've also made this a void function that mutates the array to be consistent with the way the 2D one works, but outputting an array of vertices which we then concatenate to a big list would also work.

I've also added a `first` property here because I assume for 2D mode, each `vertex` may need to know if it needs to `moveTo` or if it needs to `lineTo` based on whether it's the start of a path/contour.

To support contours, if we have:
```ts
type CompositePath = Array<Segment>
```

...then a full shape would be:
```ts
type Shape = Array<CompositePath>
```

where `beginShape()` starts off the shape as `[[]]` (just one composite path, initially empty), and each `beginContour` call pushes a new empty array to the shape to add a new contour. (The main shape and contours can be treated the same.)

**Comment by GregStanton** [comment]
Replying to https://github.com/processing/p5.js/issues/6560#issuecomment-1821679406 by @davepagurek:

Thanks! This is very helpful. There's some tension between the idea of vertex classes and segment classes; I think you were right to switch over to segment classes. The deciding factor for me is that it seems natural to put all the Catmull-Rom vertices in a single object, and it doesn't make sense to think of that as a singular "vertex" object.

Below, I propose a few tweaks that combine (and hopefully refine) the interfaces we each suggested. I'm using a TypeScript-like syntax as an informal shorthand (e.g. `get` doesn't actually work in TypeScript interfaces, as far as I know).

### Proposed interfaces
If these sound good, we can test them by incorporating them into a new version of the proof of concept. (That's good because I don't have the extra time right now to think this through completely! I think it makes sense, though.)

```
interface PathSegment {
  data: Array<number | string>;
  type: string;
  get endVertex(): p5.Vector;
  addToCanvasPath(ctx: CanvasRenderingContext2D): void;
  addToVertexPath(shape: Shape, curveDetail: number): void;
}

interface Path = {
  segments: Array<PathSegment>;
  currentVertex: p5.Vector;
}

interface Shape {
  boundary: Path;
  vertexBoundary: Array<p5.Vector>;
  contours: Array<Path>;
  vertexContours: Array<Array<p5.Vector>>;
  currentPath: Path;
}
```

###  Notes
* `PathSegment`
     * **first:** I omitted this because I think it may be unnecessary? When endShape() loops over the segments in a new path, it knows which of the segments is first, so I think it can call moveTo() before it calls addToCanvasPath(). That's what the proof of concept does without contours, but it seems like it should work for those too.
   * **data:** Includes everything needed to specify the segment including supporting data (control points, extra vertices for Catmull-Rom splines, ellipse parameters for an arc, etc.) and an ending vertex. The starting vertex doesn't need to be stored, since that can be pulled from the currentVertex property of the relevant Path object.
   * **type:** May equal LINE, QUADRATIC, BEZIER, or CURVE (later, we can add ARC to the list). An example use is that curveVertex(x, y) can check the current segment's type, and if it's a curve, it can push x and y into its data property instead of starting a new segment.
   * **endVertex:** Can be pulled from the data property. In a Catmull-Rom spline, it's the second-to-last vertex.
   * **addToVertexPath():** Adds vertices to shape.vertexBoundary or shape.vertexContours.
 * `Path`: 
   * **Why have a property for the current vertex?** In general, it should be helpful to keep track of the ending vertex of the last segment that's been added to the canvas or vertex path; ArcSegment will definitely need it for internal calculations. 
   * **How can the current vertex be updated?** The current vertex can be updated by endShape() after it calls addToCanvasPath() or addToVertexPath().
* `Shape`
   * **Why separate boundary from contours?** Since p5 publicly distinguishes between these, separating them in the implementation has certain advantages. For example, the user can create the boundary and contours in any order, and internally, it will still be easy to distinguish the boundary from the contours.
  * **How it might work:** Internally, we can create one Shape instance. Then beginShape() can reset the shape's boundary and contours array, and it can set its currentPath to the boundary; beginContour() can push an empty path into the contours array and set currentPath to that. Then, the vertex functions don't need to know what kind of path they're working on; they just push  to the current path. Once segments have been pushed to a contour or the boundary, either endContour() or endShape() will loop over them, adding them to a canvas or vertex path; these end functions can also set the currentPath to boundary, in case the user creates the boundary after creating one or more contours.

**Comment by sudhanshuv1** [comment]
I would also like to work om this issue!

**Comment by davepagurek** [comment]
Thanks @GregStanton! I took some time to think through the implementation some more and have some more refinement to propose (or at least spelling out some of the separation of concerns more explicitly for implementers.) Ideally `p5.Renderer` just acts as an adapter between the p5 public API and `Shape`. Then, when the shape is complete, `p5.RendererGL` and `p5.Renderer2D` only need to extract the data from `Shape` in the format they need.

For easier testing and to keep the code clean, I'd ideally like to keep Shape/Path/PathSegment as decoupled as possible. There's a bit of mixing right now between what the data of these classes contain, and the algorithm of how they get constructed. Both are important, but I'd ideally like to have some invariants so that there's less confusing state management. How do you feel about these as some things to try to keep true during construction/drawing:
- Each class doesn't need to reach "up" to its parent to get info, it is all provided during construction. This is to make testing easier and so there's less for contributors to learn when fixing a bug at one level
- The data structure is the same during construction and during drawing (meaning we'd dynamically derive what part we're working on, e.g. with a `getCurrentPath()` that returns the last path in an array rather than storing it separately, so that there are less states we have to consider)

With that in mind, here are some more thoughts for **public interfaces** (omitting internal state for now as an implementation detail).

```ts
interface Shape {
  // To be called by `p5.Renderer` from public drawing API methods:
  addVertex(v: VertexInfo)
  addCurveVertex(v: VertexInfo)
  addArcVertex(v: VertexInfo, options: ArcVertexOptions)
  addQuadraticVertex(v1: VertexInfo, v2: VertexInfo)
  addBezierVertex(v1: VertexInfo, v2: VertexInfo, v3: VertexInfo)
  startPath()
  endPath()
  closePath()

  // To be called from `p5.RendererGL` and `p5.Renderer2D` to draw to the screen
  toVertices(curveDetail: number): Array<Array<VertexInfo>> // Array of arrays for each contour
  drawToCanvas(ctx: CanvasRenderingContext2D)
}

interface VertexInfo {
  position: p5.Vector

  // WebGL-specific data per vertex or control point:
  texCoord?: [number, number]
  fill?: [number, number, number, number]
}
```

Not necessary for a proof-of-concept, but just as a heads up, WebGL mode has more info than just position data for vertices, so I've made an interface for everything that might include. I think you can generally substitute `VertexInfo` everywhere we previously only had `p5.Vector` so it's not too hard of a switch when implementing things for real.

I've also only opted to include `startPath` and `endPath` here, still treating paths/contours uniformly, because while it's potentially easier for people to understand contours as being separate things from the boundary, there actually is no underlying distinction in the data: if you want to draw a venn diagram shape, you need to draw two circles, and [it doesn't matter which circle is the "contour" or the "boundary": both paths contribute to both at different times, so there's not really a reason for users to add a boundary later.](https://editor.p5js.org/davepagurek/sketches/jSbrCfT5i) For that reason I feel like it makes more sense in the renderer implementation to match how the rendering works rather than how the public API is set up. (The public API's insistence of there being one boundary if you have contours can get in the way sometimes, especially when using non-p5 data. I'd love the ability to draw shapes where you *only* have contours so that if you want to draw a font glyph from its source path information, you don't have to special-case the first contour in the glyph, and can instead treat all contours in the glyph uniformly.)

So then `Shape` has to make calls to construct the paths, and to "export" them to the canvas or to a vertex array. I imagine that means `Shape` makes new paths in the `start/endPath` methods, and otherwise forwards the vertex methods on to the paths themselves.

```ts
interface Path {
  addVertex(v: VertexInfo)
  addCurveVertex(v: VertexInfo)
  addArcVertex(v: VertexInfo, options: ArcVertexOptions)
  addQuadraticVertex(v1: VertexInfo, v2: VertexInfo)
  addBezierVertex(v1: VertexInfo, v2: VertexInfo, v3: VertexInfo)
  closePath()

  toVertices(curveDetail: number): Array<VertexInfo> // Just one array now at this level
  drawToCanvas(ctx: CanvasRenderingContext2D)

  firstVertex(): VertexInfo // For segments, if needed
}
```

The vertex adder methods on `Path` will do a lot of the heavy lifting of figuring out when to construct new segments and of what type. The idea is that that's `Path`'s responsibility, and then `Segment` is just responsible for translating that data into actual drawing commands or vertices.

Since I'm just documenting the public API, I've omitted things like the current vertex. The idea is that the path should know its own current vertex (or can be derived by getting the last vertex of its last segment), but nothing externally needs to know that.

Following @GregStanton's idea, if segments don't know their index, paths would create an array containing the first vertex of the first segment, and then calling `addToVertexPath` on each segment adds everything *except the first vertex of the segment* to the path since that will be covered by the array already. I've added a `startVertex` getter to facilitate that.

```ts
interface PathSegment {
  type: string
  index: number
  startVertex(): VertexInfo
  endVertex(): VertexInfo
  drawToCanvas(ctx: CanvasRenderingContext2D)
  addToVertexPath(shape: Shape, curveDetail: number)
}
```

**Comment by GregStanton** [comment]
Replying to https://github.com/processing/p5.js/issues/6560#issuecomment-1824900276 by @davepagurek:

Thank you for all this work! One of my favorite things about software development is iterating like this. My previous proposal needed to be adjusted to improve encapsulation, and it looks like you've taken the next steps. Instead of state changes coming from anywhere, we now have objects being responsible for modifying themselves. I see how removing internal state from the interfaces (like the `data` property in my last proposal) expresses this intent more clearly. 

Eventually, we may want to specify the interfaces, the internal state for classes that implement them, and as @limzykenneth suggested, the expected call sites for the various methods. That could be a guide for implementers. 

But first, I think I'll be able to provide better feedback on the interface details after I have a few questions answered.

### Questions
#### 2D vs. 3D
I'm now thinking we really have two shape models:

* In the 2D case, we have a segment-based model: shape &rarr; paths &rarr; path segments. The path segments are added directly to the canvas, e.g. with the native `quadraticCurveTo()` function.
* In the 3D case, we have a vertex-based model: shape &rarr; paths &rarr; vertices. The vertices are connected by line segments in the canvas.

If we follow the single-responsibility principle, it seems like these should be implemented in separate classes. People interested in the 2D functionality may think the vertex-related code is relevant to them, when in fact it's not. 

Maybe we could specify a common Shape interface that abstracts out anything common to the two cases, and then we could extend it with Shape2D and Shape3D interfaces? I'd need to spend more time to figure out what this would look like, so I figure it's good to float the general idea first.

#### Did I break it up too much?
Conceptually, our shapes are built from paths, and our paths are built from segments or vertices. It seems nice to organize our code to reflect this. On the other hand, breaking it up means a single function call turns into three. If we want to draw a shape, we now call a drawing method on the shape, and that calls a drawing method on a path, and that calls a drawing method on a segment. 

It's not currently clear to me if this is a step backward or a step forward. Maybe we should focus on breaking up the code horizontally (between 2D and 3D cases) rather than vertically (shapes, paths, and path segments)? Is it possible that doing both would mean breaking things up too much? Maybe we could just use a Path type in our specification, rather than a full interface, like you had originally?

#### Boundary vs. contours
> I'd love the ability to draw shapes where you only have contours...

I think you're right about combining "boundary" and "contours." The original idea is to make `beginShape()/endShape()` as flexible as the native canvas API by supporting composite paths. Since p5 already supports `moveTo()` operations via `beginContour()`/`endContour()`, it seems reasonable to take advantage of the full generality there as well. 

Do you see any reason why we couldn't treat `beginContour()`/`endContour()` as a general way to create paths within shapes, without requiring a separate boundary to be created first? It seems like this could be done without breaking existing sketches.

**Comment by davepagurek** [comment]
> If we follow the single-responsibility principle, it seems like these should be implemented in separate classes. People interested in the 2D functionality may think the vertex-related code is relevant to them, when in fact it's not.
> 
> Maybe we could specify a common Shape interface that abstracts out anything common to the two cases, and then we could extend it with Shape2D and Shape3D interfaces? I'd need to spend more time to figure out what this would look like, so I figure it's good to float the general idea first.

I wonder if having the ability to get vertices out of paths in general might be something we want once we have a format that makes it easy to do so, to get functionality like [`getPointAtLength()` on SVG path objects.](https://developer.mozilla.org/en-US/docs/Web/API/SVGGeometryElement/getPointAtLength) That might be overcomplicating things for now though, but maybe to have the ability to have that but also keep the vertex and canvas implementations separate, rather than my original suggestion of having methods for both on each `PathSegment`, we do something like a visitor pattern that can walk a `Path` and convert its segments into whatever format the visitor is designed to do? That would mean path data would be public, and all the segment-type-to-vertex conversions are in one place, and all the segment-type-to-canvas conversions are in another place.

> It's not currently clear to me if this is a step backward or a step forward. Maybe we should focus on breaking up the code horizontally (between 2D and 3D cases) rather than vertically (shapes, paths, and path segments)? Is it possible that doing both would mean breaking things up too much? Maybe we could just use a Path type in our specification, rather than a full interface, like you had originally?

That also works as long as we can still break out the common functionality between 2D and 3D mode. I think part of the problem right now is that because the data is handled differently across both modes, we end up with different behaviours between both. I think if the functionality to build up the path is common to both, then that fully addresses that concern for me.

Continuing to think about the visitor pattern sort of idea: if we make these as their own separate objects rather than tying it directly to `Renderer2D` and `RendererGL` (the renderers can just initialize the one they want to use), then we can still pretty easily add other operations on paths in the future.

> Do you see any reason why we couldn't treat `beginContour()`/`endContour()` as a general way to create the paths that make up a shape, without requiring a separate boundary to be created first? It seems like this could be done without breaking existing sketches.

I think that should be ok! It's currently only hard because contours are special cased in the implementation, but if we decide to treat them uniformly internally, then I think we can expose that new functionality without breaking any backwards compatibility.

**Comment by GregStanton** [comment]
Replying to https://github.com/processing/p5.js/issues/6560#issuecomment-1826355046 by @davepagurek:

Hah! I was just working on an edit to my last comment. Since you already replied, I'll add it here.

### Extensibility
What I'm about to say might be moot if we go with a different design altogether, but I wanted to get this idea on record, since the general idea is good to keep in mind.

I was thinking that, for the sake of extensibility (i.e. the open-closed principle), we might want to replace all the vertex adders&mdash; `addVertex()`, `addCurveVertex()`, etc.&mdash;with a couple general functions. For example, we might call them `addPathSegment()` and `extendPathSegment()`.

Here's a use case. Let's say someone wants to add a new type of vertex (e.g. a smooth quadratic Bezier vertex and a smooth cubic Bezier vertex, to more easily reproduce [SVG path types](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d)). Maybe they want to add to p5 itself, or maybe they want to add a new vertex type in an add-on library as @limzykenneth mentioned. Instead of having to modify the general Shape class by giving it a new vertex adder, they would just make a new user-facing vertex function and have it pass their own path segment class into  `addPathSegment()`. No changes would be necessary anywhere else.

The `extendPathSegment()` function is there in case an existing segment needs to be modified. For example, `curveVertex()` would call this function if the current path segment is a CurveSegment. I'm not sure if we'd need to allow more general modifications to the path data, but we could potentially add an `updatePathSegment()` function that gives greater write access, if that turns out to be necessary for some exotic vertex type (maybe a bit of thought could rule out this need, but I'm not sure).

At least, that's the general idea. Making this design truly extensible so that it works for add-on libraries might take some care, and I guess we might need to expose the Shape interface for that to work. But at a minimum, this should make it easier to answer the question "Where do I put my code?", which is [pretty hard to answer with the current implementation](https://github.com/processing/p5.js/issues/6459#issuecomment-1793446406).

### Visitor pattern, etc.
This looks interesting. I've been thinking about having something like getPointAtLength(). I can say more about the use case I have in mind, but I want to read over your reply more carefully first. I'll reply again after that.

**Comment by GregStanton** [comment]
Replying to https://github.com/processing/p5.js/issues/6560#issuecomment-1826355046 by @davepagurek:

### Next iteration
I quickly put together a new iteration to help us move forward. The main changes are a separation of the 2D and 3D cases, and the addition of a visitor interface for adding new functionality. Note that to add a new vertex type, it should only be necessary to (a) create a new user-facing vertex function, (b) create a new segment class in 2D and 3D, and (c) add a new method to any existing visitor classes.

_**Note:** There's a significant probability of flaws or errors in the design. There are a lot of different aspects to consider! Rather than trying to consider each aspect carefully, I decided to cobble together a rough summary (in code) of what we've discussed so far._

```
/**
BASE: 
* Shape 
* Contour
* ContourSegment
* Vertex
 */

interface Shape {
  beginContour()
  addContourSegment(contourSegment: ContourSegment) //forwarded onto Contour
  extendContourSegment(vertex: Vertex) //forwarded onto Contour
  endContour(close: string) //add CLOSE option to user-facing endContour too?
  getCurrentContour(): Contour
  getContours(): Array<Contour> //for use with getPointAtLength(), etc.?
}

interface Contour {
  addContourSegment(contourSegment: ContourSegment)
  extendContourSegment(vertex: Vertex)
  getCurrentVertex(): Vertex
  getLength() //creates getLength visitor, applies it to each segment, sums results?
  getPointsAtLength() //creates getPointsAtLength visitor, applies it to appropriate segment?
}

interface ContourSegment {
  type: string
  endVertex(): Vertex
  accept(visitor: Visitor)
}

type Vertex = Vertex2D | Vertex3D

/**
2D:
* Shape2D
* Contour2D
* ContourSegment2D
* Vertex2D
 */

interface Shape2D extends Shape {
  drawToCanvas(ctx: CanvasRenderingContext2D)
}

Interface Contour2D extends Contour {
  drawToCanvas(ctx: CanvasRenderingContext2D)
}

interface ContourSegment2D extends ContourSegment {
  drawToCanvas(ctx: CanvasRenderingContext2D)
}

type Vertex2D = p5.Vector //alias for naming consistency

/**
 3D:
 * Shape3D
 * Contour3D
 * ContourSegment3D
 * Vertex3D
 */

interface Shape3D extends Shape {
  addToVertexContour(shape: Shape, curveDetail: number)
}

Interface Contour3D extends Contour {
  toVertices(curveDetail: number): Array<Vertex3D>
}

interface ContourSegment3D extends ContourSegment {
  toVertices(curveDetail: number): Array<Vertex3D>
}

interface Vertex3D {
  position: p5.Vector

  // WebGL-specific data per vertex or control point:
  textureCoordinates?: [number, number]
  fill?: [number, number, number, number]
}

/**
VISITOR
 */
Interface Visitor {
  //2D
  visitLineSegment2D(l: LineSegment2D)
  visitQuadraticSegment2D(q: QuadraticSegment2D)
  visitBezierSegment2D(b: BezierSegment2D)
  visitCurveSegment2D(c: CurveSegment2D)
  visitArcSegment2D(a: ArcSegment2D)

  //3D
  visitLineSegment3D(l: LineSegment3D)
  visitQuadraticSegment3D(q: QuadraticSegment3D)
  visitBezierSegment3D(b: BezierSegment3D)
  visitCurveSegment3D(c: CurveSegment3D)
  visitArcSegment3D(a: ArcSegment3D)
}
```

### Questions
1. I switched the naming from "path" to "contour" (a) to avoid confusion with the native Path2D, and (b) to be consistent with the p5 naming, since we're now thinking of p5 contours as the general paths that make up a shape. Thoughts?
2. I ended up separating not just 2D and 3D, but also shapes, contours, and contour segments. This means that some function calls are forwarded down the line (e.g. when we draw a shape, the shape asks each contour to draw itself, and each contour asks its component segments to draw themselves). Can this organization be improved?
3. In Shape, I replaced `addVertex()`, etc. with the general `addContourSegment()` and `extendContourSegment()`; the adder starts a new segment and the extender extends an existing one (e.g. in the case of Catmull-Rom splines). This way, new vertex types can be added without needing to modify the Shape class. For Catmull-Rom splines, I expect the logic that determines whether to add or extend will be contained in the user-facing `curveVertex()` function rather than Contour (otherwise adding new vertex types might require modifying Contour). Does this make sense?
4. Starting/current vertices, and indices: Maybe you can help me understand this area of your proposal better, @davepagurek?
  a. For now, I removed `startVertex()` because I'm not sure how segments will come to know their start vertex; they're created with commands that never specify it (it's always assumed to be the end vertex of the last segment). If we pull a segment's start vertex from the contour it gets added to, I guess we'd have extra state to manage.
  b. Without `startVertex()`, `ArcSegment` needs another way to access its starting vertex (it needs it to convert from the parameterization of elliptical arcs that p5 will use to the one the native canvas API uses.) So, in Contour, I added getCurrentVertex() to the public API. 
  c. I also removed the index property from segments because I'm not yet sure what it's used for. 
5. In the current proposal, I just have a visitor interface for things like `getPointAtLength`, but maybe we should use visitors to convert paths to canvas commands or vertex arrays? For anyone following along, this seems to be a good introduction to the visitor pattern: https://refactoring.guru/design-patterns/visitor
6. When do we want to consider the `kind` parameter of [`beginShape()`](https://p5js.org/reference/#/p5/beginShape), or other existing features?

**Comment by GregStanton** [comment]
Hi all! 

We're going to have an online video call to hash out an approach that will allow us to accommodate contours, the 2D and 3D cases, and shape kinds; @davepagurek  and I will be there, and anyone who is interested is welcome to join!

**Date:** Thursday, December 7, 2023
**Time:** 5:00 PM Eastern Time (UTC-5)
**Place:** [BitPaper](https://bitpaper.io/)

If you're interested, just let me know, and I'll send you a link. The BitPaper platform we're using includes a shared online whiteboard, and it runs in the browser, so no installation is required. In case anyone is interested and can't make it, we'll post meeting notes here.

Thanks!

**Comment by xanderjl** [comment]
I'd love to be a fly on the wall if you'd have me @GregStanton 👀

**Comment by GregStanton** [comment]
Of course @xanderjl! Looking forward to it. I just updated the meeting announcement with the actual link to the meeting room.

**Comment by Gaurav-1306** [comment]
i would also i to join the meet.

**Comment by davepagurek** [comment]
That update looks great @GregStanton, it looks like it will handle all the situations we've though of so far!

On the call today we can go over this all again and answer any questions and clarifications other contributors might have.

**Comment by GregStanton** [comment]
### Updates from the meeting on December 7, 2023:
Thanks so much to everyone for attending! During the meeting, we adjusted the latest approach and cleared up our main questions.  I went ahead and incorporated the adjustments into [my previous comment](https://github.com/processing/p5.js/issues/6560#issuecomment-1844911703). It looks like we're ready to go ahead with the approach described there. I'll summarize the key points below. I'm filling in some details myself, so if anyone notices any errors, please let me know! 

### Summary of questions
* **Question 1:** 
  * _How will we handle shapes that require multiple calls to a vertex function before they're complete?_ 
  * Once a vertex function creates the right kind of primitive, it will ask it to add itself to the shape by calling its `addToShape()` method. This method will check the shape to see if the current primitive is of the same type; depending on what it finds, it will either extend it (by modifying its data) or it will add in the new (incomplete) primitive object. By the time `endShape()` is called and rendering occurs, the primitives will all be complete.

* **Question 2:** 
  * _The current interface only addresses the pre-rendering steps: `PrimitiveToContext2DConverter` may just add paths to the current path without stroking or filling them, and `PrimitiveToVerticesConverter` only outputs an array of vertices without rendering them. How will the rendering happen?_
  * The user-facing `endShape()` method can close the path and then call `drawShape(shape)` on the current renderer. The `drawShape()` methods will loop over the shape's primitives, produce output for rendering, and then do the actual rendering. In the 2D case, `PrimitiveToContext2DConverter` outputs path commands, and the path is then stroked and filled. In the 3D case, `PrimitiveToVerticesConverter` outputs vertices that include all necessary rendering data, and the rendering then happens according to the usual process.

* **Question 3:** 
  * _How will we want to handle `normal()` and `vertexNormal()`?_
  * These can push their data into the corresponding vertex array, since we're representing vertices as arrays of numbers.

* **Question 4:** 
  * _How likely is it that after this refactoring, we'll be able to add features that will support use cases such as reproducing and animating SVG paths (e.g. to animate the drawing of individual glyphs in typeset mathematics) or to morph one shape into another? (For example, [Manim](https://www.manim.community/) is a Python library that supports both of these features.)_
  * This seems doable! The ability to reproduce SVG paths, together with a feature like `getPointAtLength()`, should make these things possible. Some JavaScript code is already sitting around that we could use once we have the foundation in place.

### Summary of decisions
It looks like we're going to go ahead with this interface with a few adjustments.

**Adjustment 1:** Based on the discussion, it was determined that the tessellation primitives can implement the `PrimitiveShape` interface directly, without the need for an `extendByVertex()` method. The `CurveSegment` primitive will not need this type of method either. The idea is that we'll allow primitives of the same type to have write access to each other's data. I went ahead and edited the proposal above to include these changes.

**Adjustment 2:** In the 2D case, we may want to output commands to the `Path2D` interface rather than the rendering context. This may allow us to issue the basic rendering commands and any clipping commands using the same kinds of paths, since the [native canvas API's `clip()` method](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/clip) accepts a `Path2D` object.

### Next steps
Will it be helpful to break up the refactoring into smaller tasks, and then possibly assign those to individual contributors? Or would it make sense for someone to take on the whole thing? Other than this project management aspect, it seems that we're ready to start on the implementation. After that, since it will be a big change to the codebase, we'd like to use the new [visual testing system](https://github.com/processing/p5.js/issues/6576) (once it's ready), to prevent regressions.

**Comment by capGoblin** [comment]
Hi everyone, 

Apologies for the delay in updates. Here's the progress so far:

I've implemented the vertex functions for both 2D and 3D, along with visitors and contours, and initiated work on shapekinds. 

If possible I'll aim to complete shape kinds and create a PR for comprehensive review or create one with whatever done so far, soon.

**Comment by GregStanton** [comment]
Hi @capGoblin! No problem at all, and thanks for all your work! Did you see the update in the body of this issue about the new approach, the task list, and the new branch?

**Comment by GregStanton** [comment]
Pinging @Qianqianye: When I go to create new issues and link them to the task list in the body of this issue, I'd like to apply the "Good First Issue" label to at least some of them (I'll provide clear instructions to make the issues simpler). Would you be okay with giving me triage access so that I can apply labels?

I thought about asking someone else with access to apply the labels for me, but I was helping a new contributor on Discord, and I noticed that there currently aren't any issues in the p5.js repo with the "Good First Issue" label. That makes it harder for beginners to figure out where to start. If at least a few more people have triage access, there's a better chance that simpler issues will receive this label.

In general, I'm managing this refactoring issue and related issues, along with the discussions, and I've started reviewing PRs on some of the issues I started, so having triage access should be pretty helpful.

**Comment by capGoblin** [comment]
Hey @GregStanton, I overlooked the last part of the comment. That plan sounds great! In the meantime, I'll focus on addressing some other issues.

**Comment by perminder-17** [comment]
> Pinging @Qianqianye: When I go to create new issues and link them to the task list in the body of this issue, I'd like to apply the "Good First Issue" label to at least some of them (I'll provide clear instructions to make the issues simpler). Would you be okay with giving me triage access so that I can apply labels?
> 
> I thought about asking someone else with access to apply the labels for me, but I was helping a new contributor on Discord, and I noticed that there currently aren't any issues in the p5.js repo with the "Good First Issue" label. That makes it harder for beginners to figure out where to start. If at least a few more people have triage access, there's a better chance that simpler issues will receive this label.
> 
> In general, I'm managing this refactoring issue and related issues, along with the discussions, and I've started reviewing PRs on some of the issues I started, so having triage access should be pretty helpful.

I apologize for the ping, @Qianqianye. I am an active contributor to p5.js, and I've been discussing strategies on with @GregStanton on Discord to expand our organization by involving more contributors. We believe that labeling certain issues as "Good First Issue" would greatly benefit for the new contributors. I've identified several suitable "good first issues" and am willing to label them all. I'm considering two options: I could reach out to @davepagurek on Discord and request him to label the issues, but I don't want to bother him excessively as there are many. Alternatively, I could be granted triage access to label the issues independently, providing a smoother start for new contributors. I'm open to both options, and I leave the decision to you, respecting your judgment. Thankyou :)

**Comment by limzykenneth** [comment]
@GregStanton I've given you triage access to this repo.

**Comment by GregStanton** [comment]
> @GregStanton I've given you triage access to this repo.

Thanks @limzykenneth!

**Comment by limzykenneth** [comment]
@GregStanton Not sure what point you are at with this feature but I'm thinking of moving this to be a proposal in the 2.0 RFC to give it a bit more flexibility in terms of not having to consider breaking changes as much. Let me know if that sounds good to you, you can summarize what you have here as the proposal and no need to write something from scratch.

**Comment by GregStanton** [comment]
Thanks @limzykenneth! There's been some good progress so far. I've sorted out a number of the smaller tasks in the task list, and we've worked out the overall implementation. However, there are some major updates that I'd like to make, in light of the move to p5.js 2.0. 

1. I proposed [a significant API change](https://github.com/processing/p5.js/issues/5722#issuecomment-1871712771) in a related issue, based on an idea from @davepagurek. 
2. I worked out initial ideas for a minimal `p5.Shape` class that would expose some of the essential functionality we've planned. It would also make the design more extensible for add-on libraries. Dealing with this now would be good because there may be some consistency issues with `p5.Geometry` to deal with.

In short, I think this issue definitely satisfies our criteria for the 2.0 proposals, but I need to figure out how to incorporate the two updates above. Also, the comments on the current issue contain a long brainstorming discussion that led to the implementation design, but newcomers shouldn't need to read through all that. 

So, it might make sense for me to submit a clean, updated issue using the new 2.0 template. (I could tag everyone who was originally interested in this issue and link back to it for context.) Do you mind if I spend some time this week to figure out a good approach?

**Comment by GregStanton** [comment]
Okay, I'm ready for this issue to be listed as an RFC proposal, @limzykenneth. I think this should be separate from the [proposal for the new vertex API](https://github.com/processing/p5.js/issues/6766#issuecomment-1907437812), since the refactoring could be done for either API.

**Comment by GregStanton** [comment]
Some thoughts about an explicit name for the default shape kind in 2D and 3D are below. Any feedback would be great!

**Current situation:**
It looks like `TESS` is the explicit name for the default shape kind [in 3D mode](https://github.com/processing/p5.js/blob/5a7129b21da8a9dc85e47d6b8f7d066f4e770d59/src/webgl/p5.RendererGL.Immediate.js#L35), but there is no explicit name for the default shape kind [in 2D mode](https://github.com/processing/p5.js/blob/5a7129b21da8a9dc85e47d6b8f7d066f4e770d59/src/core/shape/vertex.js#L534) (it's set to `null`). The reference doesn't specify the default in either case, and the explanation for `TESS` seems unclear. 

**Solution?**
We'll want to address this as part of the refactoring, in accordance with proposal #6679. That proposal mentions `SEGMENTS` as a possible name for the default shape kind in 2D and 3D (we would eliminate `TESS`). However, I'm now thinking this name doesn't quite work. How about `SEGMENT_CHAIN`? Although `CHAIN` would be more concise, I'd probably go with the more precise term, especially since it fits well with existing names like `TRIANGLE_FAN`, etc.

Here's why I'm proposing `SEGMENT_CHAIN` in place of `SEGMENTS`:

* In the refactoring plan, we're calling the default shape classes "segments." However, `SEGMENTS` is problematic: the existing plural shape kinds (`POINTS`, `LINES`, `TRIANGLES`, `QUADS`) are all disconnected. So, `SEGMENTS` suggests a set of disconnected segments.
* `SEGMENT` would likely cause confusion since shapes of this kind won’t usually consist of a single segment. 
* `SPLINE` would make sense. This term usually refers to polynomials that are pieced together. Arcs aren’t polynomials (eventually we'll include `arcVertex()`), but there is such a thing as non-polynomial splines. The problem is that we’re already planning to use this term for `splineVertex()` (see [#6766](https://github.com/processing/p5.js/issues/6766)).

**Comment by davepagurek** [comment]
@jules-kris, you're looking into explicit names for default values right now, let me know if you have any feedback on the name for this one!

I think `SEGMENT_CHAIN` is a good working title while we think, and we can switch the name if we think of something else. I guess in the default mode, you're defining a shape by creating the contours or tracing the outlines of shapes? both "contour" and "outline" are kind of overloaded terms though, hmm. Silhouette might be a less overloaded kind-of-synonym? And some of the other modes also do that, e.g. `TRIANGLES` or `QUADS` -- maybe that's ok though and those can be thought of as special cases for when you want to draw the silhouettes of a bunch of the same kind of shape in a row?

**Comment by GregStanton** [comment]
@davepagurek: Thanks Dave! I appreciate the brainstorm! 

My initial feeling about "outline" and "silhouette" is that these both seem to refer to marking the boundary of a 2D shape, which is typically a closed curve. To be fair, `beginShape()` does use "shape" to include non-closed curves... But I'm not sure, especially if we want the same default for both 2D and 3D. For example, we might draw [a helix](https://editor.p5js.org/highermathnotes/sketches/9YTFMQmb6) in 3D mode; it's hard for me to see that as an outline or a silhouette. Also, `silhouette` is hard to spell.

But it's great to come up with lots of ideas. The more the better. I'll let your ideas percolate in my brain and see if they inspire something new!

**Comment by GregStanton** [comment]
Oh wait. How about `PATH`?

**Comment by davepagurek** [comment]
Path sounds pretty good to me, and it's definitely easier to spell! I suppose its a little overloaded in the graphics space because e.g. in SVGs and in the native canvas context, they have objects called paths rather than modes, but p5 doesn't have any objects or other functions using that name, and this shape mode works pretty similarly to the SVG/native canvas path drawing APIs.

**Comment by GregStanton** [comment]
Thanks @davepagurek! Good point about objects vs. modes. Hopefully that doesn't cause too much confusion. For what it's worth, we will have path objects under the hood (`LinePath`, `BezierPath`, `SplinePath`), although I don't expect we'll expose those to the user, at least not for now.

**Comment by GregStanton** [comment]
**One more idea for the name of the default shape kind:**

If a future version of p5 were to allow all shape kinds to work with all vertices, then we'd be forced to interpret `LINES` more generously. In some contexts, "line" is indeed used to mean "curve" (we have more precise terms "rectilinear" and "curvilinear" for this reason). Then `LINE` might work even better than `PATH` as the name of the default shape kind. I thought I should add this observation here, for the record, but I'm not sure yet which name is better.

**Comment by GregStanton** [comment]
**Update:** After a quick real-time discussion with @davepagurek , it looks like `PATH` and `LINE` are our top candidates for the name of the default shape kind. We put a little thought into this, since users may sometimes need to explicitly name the default shape kind, due to the new way contours will work.

@davepagurek: After writing down our observations (see below), I'm now leaning toward `PATH` rather than `LINE`. Do you have a preference?

@jules-kris: If you have any feedback, that would be great! If you want to skip the foregoing discussion and just focus on this comment, it should summarize the important things.

**General criteria:**
It'd be nice to fulfill as many of the criteria below as we can (I also adapted these into [a general list of criteria](https://github.com/processing/p5.js/issues/6679#issuecomment-2452709150) for #6679).

1. _Descriptiveness_
   - Implies straight, curved, or composite
   - Implies open or closed
   - Implies planar (fits in a 2D sketch) or non-planar (requires a 3D sketch)
2. _Consistency_
   - Is free of conflicts with names of other shape kinds
   - Is free of conflicts with p5 API in general
   - Is free of conflicts with other APIs (mainly the Canvas and SVG APIs)
   - Is free of conflicts in related contexts (art, math, ...)
3. _Simplicity_
   - Familiar to beginners
   - Easy to spell
   - Concise
4. _Durability_ (in the future, its scores on the other criteria will not be reduced)

**Assessment of `PATH`:**

* _Descriptiveness_: Excellent
* _Consistency_: Good 
* _Simplicity_: Excellent
* _Durability_: Good

Notes:
1. The consistency would be excellent, except that Canvas and SVG paths are user-facing objects, rather than modes. For now at least, p5's path objects will just be internal. (Specifically, a path would be implemented as a contour object with the `PATH` kind.) 
2. The durability would be excellent, except that the name `LINE` would be more consistent with `LINES` in the future, if p5 were to allow all shape kinds to work with all vertex types.

**Assessment of `LINE`:**

* _Descriptiveness_: Average
* _Consistency_: Average
* _Simplicity_: Excellent
* _Durability_: Excellent

Notes:
1. Descriptiveness isn't "Excellent" since "line" is often used to specifically mean a straight line, even though "line" is used to refer to straight or curved lines in some contexts, including in math and in art. As @davepagurek noted in conversation, it also may imply that it makes only stroked shapes, as in [line art](https://en.wikipedia.org/wiki/Line_art).
2. Consistency is "Average" since `LINE` conflicts with the meaning of `line()`, and since it conflicts with internal class names (at least currently, and resolving these conflicts might require longer class names).
3. Durability is "Excellent" rather than "Good" because it would fit well with the existing `LINES` name if all shape kinds eventually work with all vertex types.

**Plan for now:**
I'll use `PATH` for now. We can potentially change this up until the 2.0 release.

**Comment by golanlevin** [comment]
I was about to file a new issue, then found this one. 
As you are aware, mixed vertex types are not playing well together. I'm currently working on an SVG-export library specifically for pen-plotting, to which this issue is directly relevant. It seems like y'all are far ahead on thinking about this, so I'm just parking this here as an illustration of the current state of things (v.1.11.0) for mixed vertex types. Thanks for all of your effort on this. 

<img width="796" alt="problems-with-mixed-vertices" src="https://github.com/user-attachments/assets/324abe4f-1822-4bc1-a4ed-2d540450bb5a">

A live version of this demo is here: 
[https://editor.p5js.org/golan/sketches/gGBFMTvaI](https://editor.p5js.org/golan/sketches/gGBFMTvaI)

```
let correctImage;

function setup() {
  createCanvas(800, 500);
  correctImage = loadImage("correct.png"); 
}

function draw() {
  background(220);
  
  text("Mixed vertex types don't play well together:", 75, 70); 
  
  push();
  translate(0,0); 
  text("Bezier only", 75, 90); 
  beginShape();
  vertex( 75, 100); 
  vertex( 80, 100); 
  bezierVertex(100,100, 110,130, 125,130);
  bezierVertex(140,130, 150,100, 170,100);
  vertex(175, 100);
  vertex(175, 125);
  vertex(200, 150);
  vertex(175, 175);
  vertex(175, 200);
  vertex(150, 205);
  // quadraticVertex(125,140, 100,195); // commented out
  vertex(100, 195);
  vertex( 75, 200);
  vertex( 75, 175); 
  vertex( 75, 125); 
  endShape(CLOSE); 
  pop(); 
  
  push();
  translate(150,0); 
  text("Quadratic only", 75, 90); 
  beginShape();
  vertex( 75, 100); 
  vertex( 80, 100); 
  //bezierVertex(100,100, 110,130, 125,130); // commented out
  //bezierVertex(140,130, 150,100, 170,100); // commented out
  vertex(175, 100);
  vertex(175, 125);
  vertex(200, 150);
  vertex(175, 175);
  vertex(175, 200);
  vertex(150, 205);
  quadraticVertex(125,140, 100,195); 
  vertex(100, 195);
  vertex( 75, 200);
  vertex( 75, 175); 
  vertex( 75, 125); 
  endShape(CLOSE); 
  pop(); 

  push();
  translate(300,0); 
  text("Bezier + Quadratic (Wrong)", 75, 90); 
  beginShape();
  vertex( 75, 100); 
  vertex( 80, 100); 
  bezierVertex(100,100, 110,130, 125,130);
  bezierVertex(140,130, 150,100, 170,100);
  vertex(175, 100);
  vertex(175, 125);
  vertex(200, 150);
  vertex(175, 175);
  vertex(175, 200);
  vertex(150, 205);
  quadraticVertex(125,140, 100,195); 
  vertex(100, 195);
  vertex( 75, 200);
  vertex( 75, 175); 
  vertex( 75, 125); 
  endShape(CLOSE); 
  pop(); 
  
  push();
  translate(500,0); 
  text("Bezier + Quadratic (Correct)", 75, 90);
  image(correctImage, 75,100); 
  pop(); 
  
  
  
  text("Adding a curveVertex alters the entire shape, but shouldn't:", 75, 270); 
  
  push();
  translate(0,200); 
  text("2 vertexes on left", 75, 90); 
  beginShape();
  vertex( 75, 100); 
  vertex( 80, 100); 
  vertex(175, 100);
  vertex(175, 125);
  vertex(200, 150);
  vertex(175, 175);
  vertex(175, 200);
  vertex(150, 205);
  vertex(100, 195);
  vertex( 75, 200);
  vertex( 75, 175); 
  vertex( 60, 160); // regular vertex added
  vertex( 90, 140); // regular vertex added
  vertex( 75, 125); 
  endShape(CLOSE); 
  pop(); 
  
  push();
  translate(150,200); 
  text("2 curveVertexes on left instead (Wrong)", 75, 90); 
  beginShape();
  vertex( 75, 100); 
  vertex( 80, 100); 
  vertex(175, 100);
  vertex(175, 125);
  vertex(200, 150);
  vertex(175, 175);
  vertex(175, 200);
  vertex(150, 205);
  vertex(100, 195);
  vertex( 75, 200);
  vertex( 75, 175); 
  curveVertex( 60, 160); // curveVertex added instead
  curveVertex( 90, 140); // curveVertex added instead
  vertex( 75, 125); 
  endShape(CLOSE); 
  pop(); 
}
```

**Comment by GregStanton** [comment]
Thanks for sharing @golanlevin! It's great to hear that this project may help with the library you're working on! Would you be interested in doing some user testing with your library once everything is ready? If so, you may want to check out the [new vertex-function API](https://github.com/processing/p5.js/issues/6766) for p5.js 2.0, since it looks like we're going to be switching to that.

**Comment by golanlevin** [comment]
@GregStanton Thanks for pointing me to the new vertex API for p5 v.2!
The SVG-export library I'm making is specifically intended for p5.js, so this is directly relevant to me. I appreciate that there will be some breaking changes, and I'll check the roadmap closely!

**Comment by GregStanton** [comment]
@golanlevin Sure thing! I'm glad that was helpful :)

**Comment by GregStanton** [comment]
@golanlevin Wow, looks cool! The README is quite helpful.

I'm glad to hear you like the new vertex-function API in #6766! Would you mind adding a comment with your feedback on that issue? It's nice to have as much community feedback as possible, and it helps to be able to see it in one place. A sentence would be fine :)

We're actively working toward the new API in the [dev-2.0 branch](https://github.com/processing/p5.js/tree/dev-2.0), but we're not there yet. Some detailed aspects of the design (e.g. for splines) are still being discussed.

For a general sense of what all is being worked on, there's a [public project tracker for p5.js 2.0](https://github.com/orgs/processing/projects/21), as you may have noticed. I'm not sure if the timeline is public yet, but [the plan](https://github.com/processing/p5.js/issues/7261) has been to have something working by the end of 2024. 

I think @davepagurek will know whether there's a more specific date we can comment on. He may also have a better idea about a compatibility module. (I think the idea is to continue supporting old features in a compatibility add-on, for some time after 2.0 is released.)

**Comment by limzykenneth** [comment]
@golanlevin Had a quick glanced at the code of your library and I think for the most part there probably won't be breaking changes and if there are, should be easy to fix. 

I plan to go through at least some of the addon libraries at a later stage to help start the transition of addon libraries to use newer syntax, I can take a close look at yours at that stage as well if you like.

**Comment by davepagurek** [comment]
This will be in 2.0 as of https://github.com/processing/p5.js/pull/7373!
